### DISCUSSION TITLE
Removing client, Subscribe topic is empty.

### DISCUSSION THREAD

**Discussion by sivaNbalusu5** [halfgaar/FlashMQ#37] [Q&A] [answered]
Hi,

I have a consumer running in aync loop and as long as there are new messages published to the topic(using another publish client running in while loop ), I don't see any issues but if the messages are not available for few seconds(using another publish client running in for loop ) I see "Operation timed out" in 12 seconds  after publisher is done. I have tried setting up keepalive parameter to 600s but i still see timeout happening in 12 seconds. @halfgaar

I am using reconnect example from: https://pypi.org/project/asyncio-mqtt/ 

[2023-03-08 19:59:24] [NOTICE] Client '[ClientID='taFZMGPgNREbVLnAQWIMYzn', username='', fd=74, keepalive=600s, transport='TCP/SSL', address='10.105.0.1', prot=5.0, clean=1]' logged in successfully
[2023-03-08 19:59:24] [NOTICE] Removing client '[ClientID='taFZMGPgNREbVLnAQWIMYzn', username='', fd=74, keepalive=600s, transport='TCP/SSL', address='10.105.0.1', prot=5.0, clean=1]'. Reason(s): Subscribe topic is empty.
[2023-03-08 19:59:36] [NOTICE] Accepting connection from: address='10.105.0.1', transport='TCP/SSL', fd=74
[2023-03-08 19:59:36] [NOTICE] Client '[ClientID='6FklpDM4KkyVSCDK3rK7pFc', username='', fd=74, keepalive=600s, transport='TCP/SSL', address='10.105.0.1', prot=5.0, clean=1]' logged in successfully
[2023-03-08 19:59:36] [NOTICE] Removing client '[ClientID='6FklpDM4KkyVSCDK3rK7pFc', username='', fd=74, keepalive=600s, transport='TCP/SSL', address='10.105.0.1', prot=5.0, clean=1]'. Reason(s): Subscribe topic is empty.
[2023-03-08 19:59:48] [NOTICE] Accepting connection from: address='10.105.0.1', transport='TCP/SSL', fd=74
[2023-03-08 19:59:48] [NOTICE] Client '[ClientID='KArFWCJw3Csns2Ho5QXGpph', username='', fd=74, keepalive=600s, transport='TCP/SSL', address='10.105.0.1', prot=5.0, clean=1]' logged in successfully
[2023-03-08 19:59:48] [NOTICE] Removing client '[ClientID='KArFWCJw3Csns2Ho5QXGpph', username='', fd=74, keepalive=600s, transport='TCP/SSL', address='10.105.0.1', prot=5.0, clean=1]'. Reason(s): Subscribe topic is empty.

**Comment by halfgaar** [comment]
Weird. The message "Subscribe topic is empty" means a client subscribes with an empty topic. This spec says this is a protocol violation and therefore the client should be disconnected. I'd like to see the actual packet received. But, so far I am unable to reproduce. I have running as consumer:

```
import asyncio
import asyncio_mqtt as aiomqtt
import ssl

tls_params = aiomqtt.TLSParameters(
    ca_certs=None,
    certfile=None,
    keyfile=None,
    cert_reqs=ssl.CERT_NONE,
    tls_version=ssl.PROTOCOL_TLS,
    ciphers=None,
)

async def main():
    reconnect_interval = 5  # In seconds
    while True:
        try:
            async with aiomqtt.Client("192.168.178.127", port=8883, tls_params=tls_params) as client:
                async with client.messages() as messages:
                    await client.subscribe("#")
                    async for message in messages:
                        print(message.payload.decode())
        except aiomqtt.MqttError as error:
            print(f'Error "{error}". Reconnecting in {reconnect_interval} seconds.')
            await asyncio.sleep(reconnect_interval)



asyncio.run(main())
```

And as producer: `( while true; do date; sleep 1; done) | mosquitto_pub -t 'afd' -h 192.168.178.127 -l`

It can turn it off or on as much as I like.

Can you provide your exact example?

**Reply by sivaNbalusu5** [reply]
@halfgaar I think context manager is the problem here, The below tasks are gathered and run using use_uvloop. Please find the code below if it helps. Also can you please recommend any asyncio mqtt client in python? 


```
    @retry(stop=stop_after_attempt(2), wait=wait_fixed(2))
    async def start(self):
        try:
            loop = asyncio.get_event_loop()
            print(f"hostname: {self.hostname}")
            print(f"Port: {self.port}")
            print(f"topic: {self.topic}")
            print(self.tls_params)
            self.__consumer = aiomqtt.Client(hostname=str(self.hostname), port=int(self.port), tls_params=self.tls_params, **self.consumer_args)
        except aiomqtt.MqttError as e:
            self.logger.error(f"error starting consumer, retrying {str(e)}")
            raise e

    async def __anext__(self):
        print("here 3")
        if self.codec.multipart:
            packets = int(await self._message())
            x = [await self._message() for _ in range(packets)]
        else:
            x = await self._message()
        return self.codec.decode(x)

    async def _message(self):
        reconnect_interval = 2
        while True:
            try:
                async with self.__consumer as client:
                    await client.subscribe(str(self.topic))
                    async with client.messages() as messages:
                        async for message in messages:
                            return message.payload
            except aiomqtt.MqttError as e:
                print(datetime.datetime.now().isoformat())
                self.logger.error("error is from consumer: %s", e)
                self.logger.info("re-initializing kafka consumer")
                print(datetime.datetime.now().isoformat())
                await asyncio.sleep(reconnect_interval)
```

**Reply by halfgaar** [reply]
Can you post your actual complete example, including imports and dependencies?

**Reply by sivaNbalusu5** [reply]
[consumer.py.zip](https://github.com/halfgaar/FlashMQ/files/10935926/consumer.py.zip)
and Consumer is a call with just one abstract method
@abc.abstractmethod
    async def __anext__(self):

**Reply by sivaNbalusu5** [reply]
@halfgaar

**Reply by halfgaar** [reply]
Sorry, I still don't know how to run this. First, apparently I had to `pip install tenacity`, which I did. Then the import `from consumer.base import Consumer` fails, so I just remove it, but then the code does nothing? It doesn't have an entrypoint. I tried to figure out what you meant with `async def anext(self)`, but then I saw when editing your comment you didn't use a code block and tried to type `__anext__`.

I'd like to help you figure if this is a problem in FlashMQ, but I really do need a workable reproduction without having to guess.

**Reply by sivaNbalusu5** [reply]
@halfgaar its a vast library to paste here. Context manager concept is not syncing up with the project structure. I am able to use another library and solve the issue. Thanks for the quick reply :)

https://amqtt.readthedocs.io/en/latest/references/mqttclient.html

**Reply by halfgaar** [reply]
If it's really the case that under certain conditions the client can send an empty topic as subscribe path, that would be a bug in that client. I was going to offer you could to connect to `demo.flashmq.org` after I prepared some debug stuff, but if you got it running, then I guess you're good?

**Reply by sivaNbalusu5** [reply]
@halfgaar Yes i am good for now. demo.flashmq.org option looks good as well because amqtt supports mqtt3.1.1 only where as the previous one supports mqtt5 and has way more activity from the community.

**Reply by halfgaar** [reply]
The server at demo.flashmq.org is now running a debug build that writes the mqtt packet out to disk before it throws that error. If you can reproduce it there, that would be great.

**Reply by sivaNbalusu5** [reply]
@halfgaar I have started the job, let me know whats happening on the other side.

**Reply by halfgaar** [reply]
@sivaNbalusu5  I saw your client, but nothing out of the ordinary happened. I did notice you're not connecting with SSL?

**Comment by sivaNbalusu5** [comment]
@halfgaar I need ca cert for demo.flashmq.org to connect via ssl.

**Reply by halfgaar** [reply]
@sivaNbalusu5  You can download it here: https://www.flashmq.org/demo/

**Reply by sivaNbalusu5** [reply]
@halfgaar I am seeing this error when i use the cert from  https://www.flashmq.org/demo/


EOF occurred in violation of protocol (_ssl.c:1129)

here are the parameters I am using:

tls_params = aiomqtt.TLSParameters(
                ca_certs="flashmq-ca.crt",
                certfile=None,
                keyfile=None,
                cert_reqs=ssl.CERT_REQUIRED,
                tls_version=ssl.PROTOCOL_TLSv1_2,
                ciphers=None,
            )

**Reply by halfgaar** [reply]
I keep seeing this:

```
[2023-03-12 21:59:30] [NOTICE] Accepting connection from: address='1.2.3.4 (redacted)', transport='TCP/Non-SSL', fd=25                                                                                                                                                                
[2023-03-12 21:59:30] [ERROR] Protocol error: Invalid variable header length. Garbage?. Removing client. 
```

It says non-ssl. Are you connecting to the right port?

**Reply by sivaNbalusu5** [reply]
I am using 

hostname: demo.flashmq.org
Port: 8883
topic: test/test1

**Reply by halfgaar** [reply]
For your IP I also see this actually:

```
[2023-03-13 05:30:17] [NOTICE] Accepting connection from: address='1.2.3.4', transport='TCP/SSL', fd=25
[2023-03-13 05:30:17] [NOTICE] Client '[ClientID='siva-test123', username='', fd=25, keepalive=59s, transport='TCP/SSL', address='1.2.3.4', prot=3.1.1, clean=1]' logged in successfully
[2023-03-13 05:30:28] [NOTICE] Removing client '[ClientID='siva-test123', username='', fd=25, keepalive=59s, transport='TCP/SSL', address='1.2.3.4', prot=3.1.1, clean=1]'. Reason(s): socket disconnect detected
```

But there are no errors. The string 'ubscribe topic is empty' does not appear anywhere in the logs.

I also note it's mqtt 3.1.1?

**Reply by sivaNbalusu5** [reply]
@halfgaar Please ignore that log, I am currently testing two things and this log belongs to this guy:

https://github.com/Yakifo/amqtt/discussions/146

**Reply by halfgaar** [reply]
@sivaNbalusu5 you'll do them a big favor if you put code in tripe backtick blocks. The way it's now, all formatting is gone, and accidentally subject to markdown interpretation.

I now see the problem, coming from IP `23.29.x.y`, from an mqtt5 client. The subscribe packet is:

```
00000000  82 10 00 d9 00 00 0a 74  65 73 74 2f 74 65 73 74  |.......test/test|
00000010  31 00                                             |1.|
00000012
```

Let's break it down:

* 0x82 = 10000010 = subscribe
* 0x10 = 00010000 = 16 = remaining length 16 bytes
* 0x00 and 0xd9 = packet id = 217. I see this incrementing packet by packet
* 0x00 = property length
* 0x00 0x0a = topic length = 10. `test/test1` is 10 chars.
* 0x00 = subscription options = QoS 0.

This looks like a correct packet. I'll investigate more.

**Reply by sivaNbalusu5** [reply]
@halfgaar I am able to reproduce this, please ignore the above message (EOF occurred in violation of protocol (_ssl.c:1129). 


"Operation timed out" was the original log on the client side.

**Reply by halfgaar** [reply]
I see you stopped your client. I was close with a debugger, but now I can't continue.

**Reply by halfgaar** [reply]
I found the problem @sivaNbalusu5. I pushed the branch https://github.com/halfgaar/FlashMQ/tree/fix-empty-subscribe-topic

Can you build from source? I'm going to release this soon, but another confirmation would be nice.

**Reply by sivaNbalusu5** [reply]
@halfgaar sorry the connection was broken and I haven't paid attention. running it now!

**Reply by halfgaar** [reply]
@sivaNbalusu5 didn't you see my last message? I already fixed it.

**Reply by sivaNbalusu5** [reply]
@halfgaar I have seen it , already built the docker image, deployed and I am about to test it :) I started the client just in case if you want to debug more. 

I will post more updates soon.

**Reply by halfgaar** [reply]
I'm curious to hear if you have test results :)

**Comment by sivaNbalusu5** [comment]
@halfgaar I am OOO last week, The preliminary test looks good, I will do more testing today.

**Reply by halfgaar** [reply, marked as answer]
I already released the fix. I had faith in it :)

**Comment by halfgaar** [comment]
Fix was released.

**Reply by sivaNbalusu5** [reply]
@halfgaar Thanks and everything looks stable.
